### PR TITLE
ISPN-6306 Dist Stream iterator can return with no results in some cases

### PR DESCRIPTION
**Pull request by wburns**
* Closing timing issue by checking queue again after completion

https://issues.jboss.org/browse/ISPN-6335

### CODE REVIEW

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java [hunk 1]
```diff
@@ -733,13 +733,22 @@ public R get() {
...
                }
                return null;
             }
             nextLock.lock();
             try {
                boolean interrupted = false;
-               while ((entry = queue.poll()) == null && !completed.get()) {
+               // We have to poll queue after completed just in case if we completed and sent an item to the
+               // queue and completed iterator that we don't lose the item
+               while ((entry = queue.poll()) == null && !completed.get() && (entry = queue.poll()) == null) {
```
Comment (anistor) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: poll twice in a row?

Comment (anistor) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: I understand the reason for it but was wondering if there's maybe a way to write it in a less weird way than relying on short-circuit evaluation.

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java [hunk 2]
```diff
@@ -733,13 +733,22 @@ public R get() {
             if (completed.get()) {
```
Comment (anistor) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: I think I saw some unused imports in DistributedCacheStream.

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java (hunk 1 above)
Comment (danberindei) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: +1, one assignment in a `while` condition is already too much :)

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java [hunk 3]
```diff
@@ -733,13 +733,22 @@ public R get() {
...
+                  return entry;
                }
                return null;
             }
             nextLock.lock();
             try {
                boolean interrupted = false;
-               while ((entry = queue.poll()) == null && !completed.get()) {
+               // We have to poll queue after completed just in case if we completed and sent an item to the
+               // queue and completed iterator that we don't lose the item
```
Comment (danberindei) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: Could you clarify this a bit? I'm still not sure what's going on with the completed iterator :)

Comment (wburns) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: I tried to explain this in more detail at https://issues.jboss.org/browse/ISPN-6335.  Did you want me to move this over to here?

Comment (wburns) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: Also am adding some more wording to describe it.

File: core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java (hunk 2 above)
Comment (wburns) [inline core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java]: Removed Configuration from imports.

**Comment by wburns** [comment]
Updated, hopefully it is better :)

**Comment by danberindei** [comment]
Integrated, thanks Will!

**Pull request closed without merging**